source/sets_functions_and_relations.rst

```diff
@@ -1293,4 +550,457 @@ at an appropriate point in the proof.
+    by { ext, refl }
+
+
+If ``s`` is a set of elements of type ``α``,
+the library also defines ``image f s``,
+written ``f '' s``,
+to be ``{y | ∃ x, x ∈ s ∧ f x = y}``.
+So a hypothesis  ``y ∈ f '' s`` decomposes to a triple
+``⟨x, xs, xeq⟩`` with ``x : α`` satisfying the hypotheses ``xs : x ∈ s``
+and ``xeq : f x = y``.
+The ``rfl`` tag in the ``rintros`` tactic (see :numref:`the_existential_quantifier`) was made precisely
+for this sort of situation.
+
+.. code-block:: lean
+
+    import data.set.function
+
+    variables {α β : Type*}
+    variable  f : α → β
+    variables s t : set α
+
+    -- BEGIN
+    example : f '' (s ∪ t) = f '' s ∪ f '' t :=
+    begin
+      ext y, split,
+      { rintros ⟨x, xs | xt, rfl⟩,
+        { left, use [x, xs] },
+        right, use [x, xt] },
+      rintros (⟨x, xs, rfl⟩ | ⟨x, xt, rfl⟩),
+      { use [x, or.inl xs] },
+      use [x, or.inr xt]
+    end
+    -- END
+
+Notice also that the ``use`` tactic applies ``refl``
+to close goals when it can.
+
+Here is another example:
+
+.. code-block:: lean
+
+    import data.set.function
+
+    variables {α β : Type*}
+    variable  f : α → β
+    variables s t : set α
+
+    -- BEGIN
+    example : s ⊆ f ⁻¹' (f '' s) :=
+    begin
+      intros x xs,
+      show f x ∈ f '' s,
+      use [x, xs]
+    end
+    -- END
+
+We can replace the line ``use [x, xs]`` by
+``apply mem_image_of_mem f xs`` if we want to
+use a theorem specifically designed for that purpose.
+But knowing that the image is defined in terms
+of an existential quantifier is often convenient.
+
+Here is a long list of set-theoretic identities for
+you to enjoy.
+You don't have to do all of them at once;
+do a few of them,
+and set the rest aside for a rainy day.
+
+.. code-block:: lean
+
+    import data.set.function
+
+    open set function
+
+    variables {α β : Type*}
+    variable  f : α → β
+    variables s t : set α
+    variables u v : set β
+
+    -- BEGIN
+    example (h : injective f) : f ⁻¹' (f '' s) ⊆ s :=
+    sorry
+
+    example : f '' (f⁻¹' u) ⊆ u :=
+    sorry
+
+    example (h : surjective f) : u ⊆ f '' (f⁻¹' u) :=
+    sorry
+
+    example (h : s ⊆ t) : f '' s ⊆ f '' t :=
+    sorry
+
+    example (h : u ⊆ v) : f ⁻¹' u ⊆ f ⁻¹' v :=
+    sorry
+
+    example : f ⁻¹' (u ∪ v) = f ⁻¹' u ∪ f ⁻¹' v :=
+    sorry
+
+    example : f '' (s ∩ t) ⊆ f '' s ∩ f '' t :=
+    sorry
+
+    example (h : injective f) : f '' s ∩ f '' t ⊆ f '' (s ∩ t) :=
+    sorry
+
+    example : f '' s \ f '' t ⊆ f '' (s \ t) :=
+    sorry
+
+    example : f ⁻¹' u \ f ⁻¹' v ⊆ f ⁻¹' (u \ v) :=
+    sorry
+
+    example : f '' s ∩ v = f '' (s ∩ f ⁻¹' v) :=
+    sorry
+
+    example : f '' (s ∩ f ⁻¹' u) ⊆ f '' s ∪ u :=
+    sorry
+
+    example : s ∩ f ⁻¹' u ⊆ f ⁻¹' (f '' s ∩ u) :=
+    sorry
+
+    example : s ∪ f ⁻¹' u ⊆ f ⁻¹' (f '' s ∪ u) :=
+    sorry
+    -- END
+
+You might also enjoy looking up for making up some
+identities involving indexed unions and proving them.
```

Comment:
That is fair. I was being lazy and worrying that the first section was already too long. I rationalized that we can't do everything, and having shown readers how to work with `Union`, they can figure out the rest. But they should at least be mentioned, and it would be good to include a few exercises. 

Since I am short of time right now, I put `TODO` comments in the text. I may not get to it before the workshop next week, but I won't forget.